synapse/cortex.py

```diff
@@ -6179,6 +6199,557 @@ async def iterTagPropRows(self, layriden, tag, prop, form=None, stortype=None, s
             yield item
 
+    def _initVaults(self):
+        self.vaultsdb = self.slab.initdb('vaults')
+        # { idenb: s_msgpack.en(vault), ... }
+
+        self.vaultsbytypedb = self.slab.initdb('vaults:bytype', dupsort=True)
+        # { vtype.encode(): idenb, ... }
+
+        self.vaultsbynamedb = self.slab.initdb('vaults:byname')
+        # { name.encode(): idenb, ... }
+
+        # TSI = type, scope, iden. This is used to deconflict uniqueness of
+        # scoped vaults without requiring a bunch of other indexes.
+        self.vaultsbyTSIdb = self.slab.initdb('vaults:byTSI')
+        # { TSI.encode(): idenb, ... }
+
+        self.vaultdefaultsdb = self.slab.initdb('vaults:defaults')
+        # { vtype.encode(): s_msgpack.en(default), ... }
+
+    def _getVaults(self):
+        genr = self.slab.scanByFull(db=self.vaultsdb)
+        for idenb, byts in genr:
+            vault = s_msgpack.un(byts)
+            vault['iden'] = s_common.ehex(idenb)
+            yield vault
+
+    def _getVaultByIden(self, iden, throw=False):
+        idenb = s_common.uhex(iden)
+        return self._getVaultByIdenb(idenb, throw=throw)
+
+    def _getVaultByIdenb(self, idenb, throw=False):
+        byts = self.slab.get(idenb, db=self.vaultsdb)
+        if byts is None:
+            if throw:
+                iden = s_common.ehex(idenb)
+                raise s_exc.NoSuchIden(mesg=f'Vault not found for iden: {iden}')
+            return None
+
+        vault = s_msgpack.un(byts)
+        vault['iden'] = s_common.ehex(idenb)
+
+        return vault
+
+    def _getVaultByTSI(self, vtype, scope, iden):
+        if scope == 'global':
+            tsi = f'{vtype}:global'
+        elif scope in ('user', 'role'):
+            tsi = f'{vtype}:{scope}:{iden}'
+        else:
+            raise s_exc.BadArg(mesg=f'Invalid scope: {scope}')
+
+        idenb = self.slab.get(tsi.encode(), db=self.vaultsbyTSIdb)
+        if idenb is None:
+            return None
+            # raise s_exc.NoSuchName(mesg=f'Vault with {vtype}:{scope}:{iden} does not exist')
+
+        return self._getVaultByIdenb(idenb)
+
+    def _setVault(self, iden, vault):
+        idenb = s_common.uhex(iden)
+        name = vault.get('name')
+        vtype = vault.get('type')
+        scope = vault.get('scope')
+
+        if scope is not None:
+            ident = vault.get('ident')
+
+            if scope == 'global':
+                tsi = f'{vtype}:global'
+            else:
+                tsi = f'{vtype}:{scope}:{ident}'
+
+            self.slab.put(tsi.encode(), idenb, db=self.vaultsbyTSIdb)
+
+        self.slab.put(idenb, s_msgpack.en(vault), db=self.vaultsdb)
+        self.slab.put(name.encode(), idenb, db=self.vaultsbynamedb)
+        self.slab.put(vtype.encode(), idenb, db=self.vaultsbytypedb)
+
+    def _getVaultByName(self, name, throw=False):
+        idenb = self.slab.get(name.encode(), db=self.vaultsbynamedb)
+        if idenb is None:
+            if throw:
```

Comment:
normally we would have a `_reqVaultByName` that would throw if the vault does not exist.